File: sunnad-ios/sunnad-ios/Features/Groups/ViewModels/GroupsViewModel.swift
import Combine
import Foundation

@MainActor
final class GroupsViewModel: ObservableObject {
    private struct SharingPipelineState {
        var desiredHabitIDs: Set<UUID>
        var committedHabitIDs: Set<UUID>
        var isInFlight: Bool
    }

    @Published private(set) var groups: [UIGroup] = []
    @Published private(set) var user: UIUserState = .guest
    @Published private(set) var habits: [UIHabit] = []
    @Published private(set) var isLoading = false
    @Published private(set) var errorMessage: String?

    private let groupsRepository: GroupsRepository
    private let logger: AnalyticsLogging
    private let analytics: AnalyticsClient
    private var sharingPipelines: [UUID: SharingPipelineState] = [:]

    init(
        groupsRepository: GroupsRepository,
        logger: AnalyticsLogging,
        analytics: AnalyticsClient
    ) {
        self.groupsRepository = groupsRepository
        self.logger = logger
        self.analytics = analytics
    }

    func load(user: UIUserState, habits: [UIHabit]) async {
        self.user = user
        self.habits = habits
        errorMessage = nil

        guard !user.isGuest else {
            groups = []
            return
        }

        await refresh()
    }

    func refresh() async {
        guard !user.isGuest else {
            groups = []
            sharingPipelines = [:]
            return
        }

        isLoading = true
        defer { isLoading = false }

        do {
            let fetchedGroups = try await groupsRepository.fetchGroups().map { group in
                var uiGroup = group.asUIGroup()
                uiGroup.sharedHabitIDs = resolvedSharedHabitIDs(
                    for: uiGroup.id,
                    fallback: uiGroup.sharedHabitIDs
                )
                return uiGroup
            }
            groups = refreshGroupProgress(for: fetchedGroups)
            syncSharingPipelines(with: groups)
            errorMessage = nil
        } catch {
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.load"))
            logger.log(.storageFailure, metadata: ["scope": "groups_refresh", "error": error.localizedDescription])
        }
    }

    func refreshGroup(groupID: UUID) async {
        guard !user.isGuest else { return }
        do {
            if var group = try await groupsRepository.refreshGroup(groupID: groupID)?.asUIGroup() {
                group.sharedHabitIDs = resolvedSharedHabitIDs(
                    for: group.id,
                    fallback: group.sharedHabitIDs
                )
                mergeOrAppend(group)
                groups = refreshGroupProgress(for: groups)
                syncSharingPipeline(for: groupID)
            } else {
                groups.removeAll(where: { $0.id == groupID })
                sharingPipelines.removeValue(forKey: groupID)
            }
            errorMessage = nil
        } catch {
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.load"))
            logger.log(.storageFailure, metadata: ["scope": "groups_refresh_group", "error": error.localizedDescription])
        }
    }

    func syncHabits(_ habits: [UIHabit]) async {
        self.habits = habits
        guard !user.isGuest else { return }
        groups = refreshGroupProgress(for: groups)
    }

    func createGroup(name: String) async {
        guard !user.isGuest else { return }
        let normalizedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedName.isEmpty else { return }

        let pendingGroup = makePendingGroup(name: normalizedName, code: "…")
        groups.append(pendingGroup)
        groups = refreshGroupProgress(for: groups)

        do {
            let group = try await groupsRepository
                .createGroup(name: normalizedName)
                .asUIGroup()
            replacePendingGroup(withID: pendingGroup.id, with: group)
            groups = refreshGroupProgress(for: groups)
            syncSharingPipeline(for: group.id)
            errorMessage = nil
            analytics.trackGroup(.created)
        } catch {
            groups.removeAll(where: { $0.id == pendingGroup.id })
            groups = refreshGroupProgress(for: groups)
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.create"))
            logger.log(.storageFailure, metadata: ["scope": "groups_create", "error": error.localizedDescription])
        }
    }

    func joinGroup(code: String) async {
        guard !user.isGuest else { return }
        let normalizedCode = code.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        guard !normalizedCode.isEmpty else { return }

        let pendingGroup = makePendingGroup(
            name: "\(L10n.t("groups.join")) \(normalizedCode)",
            code: normalizedCode
        )
        groups.append(pendingGroup)
        groups = refreshGroupProgress(for: groups)

        do {
            let group = try await groupsRepository.joinGroup(code: normalizedCode).asUIGroup()
            replacePendingGroup(withID: pendingGroup.id, with: group)
            groups = refreshGroupProgress(for: groups)
            syncSharingPipeline(for: group.id)
            errorMessage = nil
            analytics.trackGroup(.joinResult(status: "success", reason: nil))
        } catch {
            groups.removeAll(where: { $0.id == pendingGroup.id })
            groups = refreshGroupProgress(for: groups)
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.join"))
            logger.log(.storageFailure, metadata: ["scope": "groups_join", "error": error.localizedDescription])
            analytics.trackGroup(.joinResult(status: "failure", reason: "error"))
        }
    }

    func renameGroup(groupID: UUID, name: String) async {
        do {
            try await groupsRepository.renameGroup(groupID: groupID, name: name.trimmingCharacters(in: .whitespacesAndNewlines))
            await refreshGroup(groupID: groupID)
        } catch {
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.rename"))
            logger.log(.storageFailure, metadata: ["scope": "groups_rename", "error": error.localizedDescription])
        }
    }

    func setGroupJoinLock(groupID: UUID, locked: Bool) async {
        do {
            try await groupsRepository.setJoinLock(groupID: groupID, locked: locked)
            await refreshGroup(groupID: groupID)
        } catch {
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.permissions"))
            logger.log(.storageFailure, metadata: ["scope": "groups_set_lock", "error": error.localizedDescription])
        }
    }

    func rotateInviteCode(groupID: UUID) async {
        do {
            _ = try await groupsRepository.rotateInviteCode(groupID: groupID)
            await refreshGroup(groupID: groupID)
        } catch {
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.permissions"))
            logger.log(.storageFailure, metadata: ["scope": "groups_rotate_code", "error": error.localizedDescription])
        }
    }

    func updateGroupSharing(groupID: UUID, habitIDs: Set<UUID>) async {
        guard !user.isGuest else { return }
        let currentSharedHabitIDs = groups.first(where: { $0.id == groupID })?.sharedHabitIDs ?? []
        applyOptimisticSharing(groupID: groupID, habitIDs: habitIDs)
        var pipeline = sharingPipelines[groupID] ?? SharingPipelineState(
            desiredHabitIDs: habitIDs,
            committedHabitIDs: currentSharedHabitIDs,
            isInFlight: false
        )
        pipeline.desiredHabitIDs = habitIDs
        sharingPipelines[groupID] = pipeline

        await flushSharingPipeline(groupID: groupID)
    }

    func updateHabitSharing(habitID: UUID, sharedGroupIDs: Set<UUID>) async {
        guard !groups.isEmpty else { return }

        // Optimistic: update UI immediately
        var snapshotByGroup: [UUID: Set<UUID>] = [:]
        for group in groups where !group.isPending {
            snapshotByGroup[group.id] = group.sharedHabitIDs
            var habitIDs = group.sharedHabitIDs
            if sharedGroupIDs.contains(group.id) {
                habitIDs.insert(habitID)
            } else {
                habitIDs.remove(habitID)
            }
            applyOptimisticSharing(groupID: group.id, habitIDs: habitIDs)
        }

        // Persist to remote
        do {
            for group in groups where !group.isPending {
                try await groupsRepository.updateSharing(groupID: group.id, habitIDs: group.sharedHabitIDs)
            }
        } catch {
            // Rollback on failure
            for (groupID, original) in snapshotByGroup {
                applyOptimisticSharing(groupID: groupID, habitIDs: original)
            }
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.permissions"))
            logger.log(.storageFailure, metadata: ["scope": "groups_update_habit_sharing", "error": error.localizedDescription])
        }
    }

    func leaveGroup(_ groupID: UUID) async {
        guard let index = groups.firstIndex(where: { $0.id == groupID }) else {
            return
        }
        let removedGroup = groups.remove(at: index)
        sharingPipelines.removeValue(forKey: groupID)
        groups = refreshGroupProgress(for: groups)

        do {
            try await groupsRepository.leaveGroup(groupID: groupID)
            errorMessage = nil
            analytics.trackGroup(.left)
        } catch {
            groups.insert(removedGroup, at: min(index, groups.count))
            groups = refreshGroupProgress(for: groups)
            syncSharingPipeline(for: groupID)
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.permissions"))
            logger.log(.storageFailure, metadata: ["scope": "groups_leave", "error": error.localizedDescription])
        }
    }

    func deleteGroup(_ groupID: UUID) async {
        guard let index = groups.firstIndex(where: { $0.id == groupID }) else {
            return
        }
        let removedGroup = groups.remove(at: index)
        sharingPipelines.removeValue(forKey: groupID)
        groups = refreshGroupProgress(for: groups)

        do {
            try await groupsRepository.deleteGroup(groupID: groupID)
            errorMessage = nil
        } catch {
            groups.insert(removedGroup, at: min(index, groups.count))
            groups = refreshGroupProgress(for: groups)
            syncSharingPipeline(for: groupID)
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.permissions"))
            logger.log(.storageFailure, metadata: ["scope": "groups_delete", "error": error.localizedDescription])
        }
    }

    func kickMember(groupID: UUID, memberID: UUID) async {
        guard let group = groups.first(where: { $0.id == groupID }) else {
            return
        }
        guard group.ownerMemberID == group.currentUserMemberID else {
            return
        }

        do {
            try await groupsRepository.kickMember(groupID: groupID, memberUserID: memberID)
            await refreshGroup(groupID: groupID)
        } catch {
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.permissions"))
            logger.log(.storageFailure, metadata: ["scope": "groups_kick_member", "error": error.localizedDescription])
        }
    }

    func setProgressDisplayMode(groupID: UUID, mode: GroupProgressDisplayMode) async {
        guard let index = groups.firstIndex(where: { $0.id == groupID }) else { return }
        groups[index].progressDisplayMode = mode
        do {
            try await groupsRepository.setProgressDisplayMode(groupID: groupID, mode: mode)
            await refreshGroup(groupID: groupID)
        } catch {
            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.permissions"))
            logger.log(.storageFailure, metadata: ["scope": "groups_set_progress_mode", "error": error.localizedDescription])
        }
    }

    func sendNudge(groupID: UUID, memberID: UUID, habitID: UUID) async -> GroupNudgeStatus {
        let habitType = habits.first(where: { $0.id == habitID })?.isDhikr == true ? "dhikr" : "binary"
        do {
            let status = try await groupsRepository.sendNudge(
                groupID: groupID,
                toUserID: memberID,
                habitID: habitID
            )
            let statusValue: String
            switch status {
            case .delivered:
                statusValue = "delivered"
            case .alreadyDelivered:
                statusValue = "already_delivered"
            case .recipientNotRegistered:
                statusValue = "recipient_not_registered"
            case .forbidden:
                statusValue = "forbidden"
            case .error:
                statusValue = "error"
            }
            analytics.trackGroup(.nudgeResult(status: statusValue, habitType: habitType))
            return status
        } catch {
            logger.log(.storageFailure, metadata: ["scope": "groups_send_nudge", "error": error.localizedDescription])
            analytics.trackGroup(.nudgeResult(status: "error", habitType: habitType))
            return .error
        }
    }

    func currentGroupSharedHabitIDs(for groupID: UUID) -> Set<UUID>? {
        groups.first(where: { $0.id == groupID })?.sharedHabitIDs
    }

    private func mergeOrAppend(_ group: UIGroup) {
        if let index = groups.firstIndex(where: { $0.id == group.id }) {
            groups[index] = group
        } else {
            groups.append(group)
        }
    }

    private func replacePendingGroup(withID pendingID: UUID, with group: UIGroup) {
        groups.removeAll(where: { $0.id == pendingID })
        mergeOrAppend(group)
    }

    private func refreshGroupProgress(for groups: [UIGroup]) -> [UIGroup] {
        return groups.map { group in
            var mutable = group
            let myMemberID = mutable.currentUserMemberID ?? mutable.ownerMemberID
            let serverMember = mutable.members.first(where: { $0.id == myMemberID })
            let serverSharedHabitsByID = Dictionary(
                uniqueKeysWithValues: (serverMember?.sharedHabits ?? []).map { ($0.habitID, $0) }
            )
            let today = Date()
            let localSharedHabits = habits.filter { mutable.sharedHabitIDs.contains($0.id) }
            let localDueSharedHabits = localSharedHabits
                .filter { $0.isScheduled(on: today) }
                .sorted { $0.displayTitle.localizedCaseInsensitiveCompare($1.displayTitle) == .orderedAscending }
                .map {
                    UISharedHabit(
                        habitID: $0.id,
                        habitTitle: $0.displayTitle,
                        habitIconSystemName: $0.iconSystemName,
                        completedToday: $0.completedToday,
                        dueToday: true,
                        streak: $0.streak,
                        rollingCompletionPercent: serverSharedHabitsByID[$0.id]?.rollingCompletionPercent
                    )
                }
            let resolvedSharedHabits: [UISharedHabit]
            if !localSharedHabits.isEmpty || mutable.sharedHabitIDs.isEmpty {
                resolvedSharedHabits = localDueSharedHabits
            } else {
                resolvedSharedHabits = serverMember?.sharedHabits.filter(\.dueToday) ?? []
            }
            let updatedMe = UIGroupMember(
                id: myMemberID,
                userID: serverMember?.userID ?? myMemberID,
                name: user.name ?? serverMember?.name ?? L10n.t("groups.you"),
                avatarURL: user.avatarURL ?? serverMember?.avatarURL,
                completedToday: resolvedSharedHabits.count { $0.completedToday },
                totalSharedHabits: resolvedSharedHabits.count,
                sharedHabits: resolvedSharedHabits
            )

            if let index = mutable.members.firstIndex(where: { $0.id == myMemberID }) {
                mutable.members[index] = updatedMe
            } else {
                mutable.members.insert(updatedMe, at: 0)
            }

            if mutable.currentUserMemberID == nil {
                mutable.currentUserMemberID = updatedMe.id
            }

            return mutable
        }
    }

    private func applyOptimisticSharing(groupID: UUID, habitIDs: Set<UUID>) {
        guard let index = groups.firstIndex(where: { $0.id == groupID }) else {
            return
        }
        groups[index].sharedHabitIDs = habitIDs
        groups = refreshGroupProgress(for: groups)
    }

    private func makePendingGroup(name: String, code: String) -> UIGroup {
        let currentUserMember = UIGroupMember(
            id: UUID(),
            name: user.name ?? L10n.t("groups.you"),
            avatarURL: user.avatarURL,
            completedToday: 0,
            totalSharedHabits: 0,
            sharedHabits: []
        )

        return UIGroup(
            id: UUID(),
            name: name,
            code: code,
            joinLocked: false,
            members: [currentUserMember],
            sharedHabitIDs: [],
            ownerMemberID: currentUserMember.id,
            currentUserMemberID: currentUserMember.id,
            isPending: true
        )
    }

    private func syncSharingPipelines(with groups: [UIGroup]) {
        let knownGroupIDs = Set(groups.map(\.id))
        sharingPipelines = sharingPipelines.filter { knownGroupIDs.contains($0.key) }
        for group in groups {
            if let current = sharingPipelines[group.id],
               current.isInFlight || current.desiredHabitIDs != current.committedHabitIDs {
                continue
            }
            sharingPipelines[group.id] = SharingPipelineState(
                desiredHabitIDs: group.sharedHabitIDs,
                committedHabitIDs: group.sharedHabitIDs,
                isInFlight: false
            )
        }
    }

    private func syncSharingPipeline(for groupID: UUID) {
        guard let group = groups.first(where: { $0.id == groupID }) else {
            sharingPipelines.removeValue(forKey: groupID)
            return
        }
        if let current = sharingPipelines[groupID],
           current.isInFlight || current.desiredHabitIDs != current.committedHabitIDs {
            return
        }
        sharingPipelines[groupID] = SharingPipelineState(
            desiredHabitIDs: group.sharedHabitIDs,
            committedHabitIDs: group.sharedHabitIDs,
            isInFlight: false
        )
    }

    private func resolvedSharedHabitIDs(for groupID: UUID, fallback: Set<UUID>) -> Set<UUID> {
        guard let pipeline = sharingPipelines[groupID] else {
            return fallback
        }
        if pipeline.isInFlight || pipeline.desiredHabitIDs != pipeline.committedHabitIDs {
            return pipeline.desiredHabitIDs
        }
        return fallback
    }

    private func flushSharingPipeline(groupID: UUID) async {
        guard var pipeline = sharingPipelines[groupID] else {
            return
        }
        guard !pipeline.isInFlight else {
            return
        }
        guard pipeline.desiredHabitIDs != pipeline.committedHabitIDs else {
            return
        }

        let targetHabitIDs = pipeline.desiredHabitIDs
        let previousCommitted = pipeline.committedHabitIDs
        pipeline.isInFlight = true
        sharingPipelines[groupID] = pipeline

        do {
            try await groupsRepository.updateSharing(groupID: groupID, habitIDs: targetHabitIDs)
            let refreshedGroup = try await groupsRepository.refreshGroup(groupID: groupID)?.asUIGroup()

            var latest = sharingPipelines[groupID] ?? pipeline
            let committedHabitIDs = refreshedGroup?.sharedHabitIDs ?? targetHabitIDs
            let serverCorrectedSelection =
                committedHabitIDs != targetHabitIDs && latest.desiredHabitIDs == targetHabitIDs

            if serverCorrectedSelection {
                latest = SharingPipelineState(
                    desiredHabitIDs: committedHabitIDs,
                    committedHabitIDs: committedHabitIDs,
                    isInFlight: false
                )
            } else {
                latest.committedHabitIDs = committedHabitIDs
                latest.isInFlight = false
            }
            sharingPipelines[groupID] = latest

            if var refreshedGroup {
                refreshedGroup.sharedHabitIDs = resolvedSharedHabitIDs(
                    for: refreshedGroup.id,
                    fallback: refreshedGroup.sharedHabitIDs
                )
                mergeOrAppend(refreshedGroup)
                groups = refreshGroupProgress(for: groups)
            } else {
                applyOptimisticSharing(groupID: groupID, habitIDs: committedHabitIDs)
            }

            if serverCorrectedSelection {
                applyOptimisticSharing(groupID: groupID, habitIDs: committedHabitIDs)
                errorMessage = L10n.t("groups.errors.unsyncedHabits")
                return
            }

            let delta = committedHabitIDs.count - previousCommitted.count
            if delta != 0 {
                analytics.trackGroup(.sharingUpdated(delta: delta, totalShared: committedHabitIDs.count))
            }

            if latest.desiredHabitIDs != latest.committedHabitIDs {
                await flushSharingPipeline(groupID: groupID)
            }
        } catch {
            var latest = sharingPipelines[groupID] ?? pipeline
            latest.isInFlight = false
            let hasNewerDesiredState = latest.desiredHabitIDs != targetHabitIDs
            sharingPipelines[groupID] = latest

            if hasNewerDesiredState {
                await flushSharingPipeline(groupID: groupID)
                return
            }

            errorMessage = friendlyGroupError(error, fallback: L10n.t("groups.errors.sharing"))
            applyOptimisticSharing(groupID: groupID, habitIDs: latest.committedHabitIDs)
            logger.log(.storageFailure, metadata: ["scope": "groups_update_sharing", "error": error.localizedDescription])
        }
    }

    private func friendlyGroupError(_ error: Error, fallback: String) -> String {
        let message = error.localizedDescription.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !message.isEmpty else { return fallback }
        let lower = message.lowercased()
        if lower.contains("group_shared_habits_habit_id_fkey") {
            return L10n.t("groups.errors.unsyncedHabits")
        }
        if lower.contains("invalid") && lower.contains("code") {
            return L10n.t("groups.errors.invalidCode")
        }
        if lower.contains("locked") || lower.contains("closed") {
            return L10n.t("groups.errors.locked")
        }
        if lower.contains("permission") || lower.contains("forbidden") {
            return L10n.t("groups.errors.permissions")
        }
        return message
    }
}

File: sunnad-android/app/src/main/java/com/arystan/almasuly/sunnadandroid/features/groups/GroupsViewModel.kt
package com.arystan.almasuly.sunnadandroid.features.groups

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arystan.almasuly.sunnadandroid.core.model.Group
import com.arystan.almasuly.sunnadandroid.core.model.GroupNudgeStatus
import com.arystan.almasuly.sunnadandroid.core.model.GroupProgressDisplayMode
import com.arystan.almasuly.sunnadandroid.domain.repository.GroupsRepository
import com.arystan.almasuly.sunnadandroid.sync.SyncCoordinator
import com.arystan.almasuly.sunnadandroid.sync.SyncTrigger
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.UUID

data class GroupsUiState(
    val groups: List<Group> = emptyList(),
    val isLoading: Boolean = false,
    val hasLoadedOnce: Boolean = false,
    val errorMessage: String? = null
)

class GroupsViewModel(
    private val groupsRepository: GroupsRepository,
    private val syncCoordinator: SyncCoordinator
) : ViewModel() {
    private data class SharingPipelineState(
        val desiredHabitIds: Set<UUID>,
        val committedHabitIds: Set<UUID>,
        val isInFlight: Boolean
    )

    private val _state = MutableStateFlow(GroupsUiState())
    val state: StateFlow<GroupsUiState> = _state.asStateFlow()
    private val sharingPipelines = mutableMapOf<UUID, SharingPipelineState>()

    fun loadGroups() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, errorMessage = null) }
            runCatching { groupsRepository.fetchGroups() }
                .onSuccess { groups ->
                    val merged = groups.map { group ->
                        group.copy(
                            sharedHabitIds = resolvedSharedHabitIds(
                                groupId = group.id,
                                fallback = group.sharedHabitIds
                            )
                        )
                    }
                    syncSharingPipelines(merged)
                    _state.update { it.copy(groups = merged, isLoading = false, hasLoadedOnce = true) }
                }
                .onFailure { error ->
                    _state.update { it.copy(isLoading = false, hasLoadedOnce = true, errorMessage = error.localizedMessage) }
                }
        }
    }

    fun clearError() {
        _state.update { it.copy(errorMessage = null) }
    }

    fun createGroup(name: String) {
        if (name.isBlank()) return
        viewModelScope.launch {
            runCatching {
                val group = groupsRepository.createGroup(name.trim())
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
                group
            }.onSuccess {
                loadGroups()
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to create group")) }
            }
        }
    }

    fun joinGroup(code: String) {
        if (code.isBlank()) return
        viewModelScope.launch {
            runCatching {
                val group = groupsRepository.joinGroup(code.trim())
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
                group
            }.onSuccess {
                loadGroups()
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to join group")) }
            }
        }
    }

    fun toggleJoinLock(group: Group) {
        viewModelScope.launch {
            runCatching {
                groupsRepository.setJoinLock(group.id, !group.joinLocked)
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
            }.onSuccess {
                refreshGroup(group.id)
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to update group settings")) }
            }
        }
    }

    fun rotateCode(group: Group) {
        viewModelScope.launch {
            runCatching {
                groupsRepository.rotateInviteCode(group.id)
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
            }.onSuccess {
                refreshGroup(group.id)
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to rotate invite code")) }
            }
        }
    }

    fun renameGroup(group: Group, name: String) {
        if (name.isBlank()) return
        viewModelScope.launch {
            runCatching {
                groupsRepository.renameGroup(group.id, name)
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
            }.onSuccess {
                refreshGroup(group.id)
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to rename group")) }
            }
        }
    }

    fun leaveGroup(group: Group) {
        viewModelScope.launch {
            runCatching {
                groupsRepository.leaveGroup(group.id)
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
            }.onSuccess {
                sharingPipelines.remove(group.id)
                loadGroups()
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to leave group")) }
            }
        }
    }

    fun updateSharing(group: Group, habitIds: Set<UUID>) {
        viewModelScope.launch {
            val currentSharedHabitIds = _state.value.groups
                .firstOrNull { it.id == group.id }
                ?.sharedHabitIds
                ?: group.sharedHabitIds

            applyOptimisticSharing(group.id, habitIds)
            val currentPipeline = sharingPipelines[group.id]
            sharingPipelines[group.id] = SharingPipelineState(
                desiredHabitIds = habitIds,
                committedHabitIds = currentPipeline?.committedHabitIds ?: currentSharedHabitIds,
                isInFlight = currentPipeline?.isInFlight ?: false
            )

            flushSharingPipeline(group.id)
        }
    }

    fun kickMember(group: Group, memberUserId: UUID) {
        viewModelScope.launch {
            runCatching {
                groupsRepository.kickMember(group.id, memberUserId)
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
            }.onSuccess {
                refreshGroup(group.id)
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to remove member")) }
            }
        }
    }

    fun deleteGroup(group: Group) {
        viewModelScope.launch {
            runCatching {
                groupsRepository.deleteGroup(group.id)
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
            }.onSuccess {
                sharingPipelines.remove(group.id)
                loadGroups()
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to delete group")) }
            }
        }
    }

    fun setProgressDisplayMode(group: Group, mode: GroupProgressDisplayMode) {
        viewModelScope.launch {
            runCatching {
                groupsRepository.setProgressDisplayMode(group.id, mode)
                syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
            }.onSuccess {
                refreshGroup(group.id)
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to update progress mode")) }
            }
        }
    }

    fun sendNudge(groupId: UUID, toUserId: UUID, habitId: UUID, onResult: (GroupNudgeStatus) -> Unit) {
        viewModelScope.launch {
            val status = groupsRepository.sendNudge(groupId, toUserId, habitId)
            onResult(status)
        }
    }

    private suspend fun refreshGroup(groupId: UUID) {
        runCatching { groupsRepository.refreshGroup(groupId) }
            .onSuccess { refreshed ->
                if (refreshed == null) {
                    sharingPipelines.remove(groupId)
                    _state.update { state ->
                        state.copy(groups = state.groups.filterNot { it.id == groupId })
                    }
                    return
                }
                val merged = refreshed.copy(
                    sharedHabitIds = resolvedSharedHabitIds(
                        groupId = refreshed.id,
                        fallback = refreshed.sharedHabitIds
                    )
                )
                syncSharingPipelineForGroup(merged.id, merged.sharedHabitIds)
                _state.update { state ->
                    val updated = state.groups.toMutableList()
                    val index = updated.indexOfFirst { it.id == merged.id }
                    if (index >= 0) {
                        updated[index] = merged
                    } else {
                        updated.add(merged)
                    }
                    state.copy(groups = updated)
                }
            }
            .onFailure { error ->
                _state.update { it.copy(errorMessage = friendlyGroupError(error, "Failed to refresh group")) }
            }
    }

    private suspend fun flushSharingPipeline(groupId: UUID) {
        val pipeline = sharingPipelines[groupId] ?: return
        if (pipeline.isInFlight) return
        if (pipeline.desiredHabitIds == pipeline.committedHabitIds) return

        val targetHabitIds = pipeline.desiredHabitIds
        val previousCommittedIds = pipeline.committedHabitIds
        sharingPipelines[groupId] = pipeline.copy(isInFlight = true)

        runCatching {
            groupsRepository.updateSharing(groupId, targetHabitIds)
            syncCoordinator.runSyncCycle(SyncTrigger.MANUAL)
            groupsRepository.refreshGroup(groupId)
        }.onSuccess { refreshed ->
            val current = sharingPipelines[groupId]
                ?: SharingPipelineState(targetHabitIds, targetHabitIds, isInFlight = false)

            val committedIds = refreshed?.sharedHabitIds ?: targetHabitIds
            val serverCorrectedSelection =
                committedIds != targetHabitIds && current.desiredHabitIds == targetHabitIds

            sharingPipelines[groupId] = if (serverCorrectedSelection) {
                SharingPipelineState(
                    desiredHabitIds = committedIds,
                    committedHabitIds = committedIds,
                    isInFlight = false
                )
            } else {
                current.copy(
                    committedHabitIds = committedIds,
                    isInFlight = false
                )
            }

            if (refreshed != null) {
                val merged = refreshed.copy(
                    sharedHabitIds = resolvedSharedHabitIds(
                        groupId = refreshed.id,
                        fallback = refreshed.sharedHabitIds
                    )
                )
                _state.update { state ->
                    val updated = state.groups.toMutableList()
                    val index = updated.indexOfFirst { it.id == merged.id }
                    if (index >= 0) updated[index] = merged else updated.add(merged)
                    state.copy(groups = updated, errorMessage = null)
                }
            } else {
                applyOptimisticSharing(groupId, committedIds)
            }

            if (serverCorrectedSelection) {
                applyOptimisticSharing(groupId, committedIds)
                _state.update {
                    it.copy(
                        errorMessage = "Some habits are not synced yet. Please wait a moment and try sharing again."
                    )
                }
                return@onSuccess
            }

            val latest = sharingPipelines[groupId]
            if (latest != null && latest.desiredHabitIds != latest.committedHabitIds) {
                flushSharingPipeline(groupId)
            } else if (previousCommittedIds != committedIds) {
                syncSharingPipelineForGroup(groupId, committedIds)
            }
        }.onFailure { error ->
            val current = sharingPipelines[groupId] ?: pipeline
            val hasNewerDesired = current.desiredHabitIds != targetHabitIds
            sharingPipelines[groupId] = current.copy(isInFlight = false)

            if (hasNewerDesired) {
                flushSharingPipeline(groupId)
                return@onFailure
            }

            applyOptimisticSharing(groupId, current.committedHabitIds)
            _state.update {
                it.copy(errorMessage = friendlyGroupError(error, "Failed to update sharing"))
            }
        }
    }

    private fun applyOptimisticSharing(groupId: UUID, habitIds: Set<UUID>) {
        _state.update { state ->
            val updated = state.groups.map { group ->
                if (group.id == groupId) group.copy(sharedHabitIds = habitIds) else group
            }
            state.copy(groups = updated)
        }
    }

    private fun syncSharingPipelines(groups: List<Group>) {
        val knownGroupIds = groups.map { it.id }.toSet()
        sharingPipelines.keys.retainAll(knownGroupIds)
        groups.forEach { group ->
            val existing = sharingPipelines[group.id]
            if (existing == null || (!existing.isInFlight && existing.desiredHabitIds == existing.committedHabitIds)) {
                sharingPipelines[group.id] = SharingPipelineState(
                    desiredHabitIds = group.sharedHabitIds,
                    committedHabitIds = group.sharedHabitIds,
                    isInFlight = false
                )
            }
        }
    }

    private fun syncSharingPipelineForGroup(groupId: UUID, sharedHabitIds: Set<UUID>) {
        val existing = sharingPipelines[groupId]
        if (existing != null && (existing.isInFlight || existing.desiredHabitIds != existing.committedHabitIds)) {
            return
        }
        sharingPipelines[groupId] = SharingPipelineState(
            desiredHabitIds = sharedHabitIds,
            committedHabitIds = sharedHabitIds,
            isInFlight = false
        )
    }

    private fun resolvedSharedHabitIds(groupId: UUID, fallback: Set<UUID>): Set<UUID> {
        val pipeline = sharingPipelines[groupId] ?: return fallback
        return if (pipeline.isInFlight || pipeline.desiredHabitIds != pipeline.committedHabitIds) {
            pipeline.desiredHabitIds
        } else {
            fallback
        }
    }

    private fun friendlyGroupError(error: Throwable, fallback: String): String {
        val message = error.localizedMessage?.trim().orEmpty()
        if (message.isBlank()) return fallback
        val lower = message.lowercase()
        return when {
            "invalid" in lower && "code" in lower -> "Invalid invite code."
            "locked" in lower || "closed" in lower -> "This group is closed to new members."
            "already" in lower && "member" in lower -> "You are already in this group."
            "group_shared_habits_habit_id_fkey" in lower ->
                "Some habits are not synced yet. Please wait a moment and try sharing again."
            "permission" in lower || "forbidden" in lower -> "You do not have permission for this action."
            else -> message
        }
    }
}
